opensrp-reveal/src/main/java/org/smartregister/reveal/sync/LocationTaskIntentService.java

```diff
@@ -38,43 +46,56 @@ private void doSync() {
         List<Location> syncedStructures = locationServiceHelper.fetchLocationsStructures();
         List<Task> synchedTasks = taskServiceHelper.syncTasks();
 
-        (new AppExecutors()).mainThread().execute(new Runnable() {
+        Pair<Boolean, Set<String>> structureIds = checkChangeInOperationalAreaAndExtractStructureIds(syncedStructures, synchedTasks);
+
+        if (structureIds.first != null && structureIds.first) {
+            Intent intent = new Intent(STRUCTURE_TASK_SYNCHED);
+            LocalBroadcastManager.getInstance(getApplicationContext()).sendBroadcast(intent);
+        }
+
+        clientProcessEvents(structureIds.second);
+
+        new AppExecutors().mainThread().execute(new Runnable() {
             @Override
             public void run() {
                 SyncServiceJob.scheduleJobImmediately(SyncServiceJob.TAG);
             }
         });
 
-        if (hasChangesInCurrentOperationalArea(syncedStructures, synchedTasks)) {
-            Intent intent = new Intent(STRUCTURE_TASK_SYNCHED);
-            LocalBroadcastManager.getInstance(getApplicationContext()).sendBroadcast(intent);
-        }
-
     }
 
-    private boolean hasChangesInCurrentOperationalArea(List<Location> syncedStructures, List<Task> synchedTasks) {
+    private Pair<Boolean, Set<String>> checkChangeInOperationalAreaAndExtractStructureIds(List<Location> syncedStructures, List<Task> synchedTasks) {
```

Comment:
It would have been better if another function was created to extract the structure ids?

Reply:
The method has a for loop and they have been combined because of optimization. Avoiding 2 for loops